nixos/modules/services/misc/forgejo.nix

```diff
@@ -501,61 +584,15 @@ in
       # lfs_jwt_secret.
       # We have to consider this to stay compatible with older installations.
       preStart =
-        let
-          runConfig = "${cfg.customDir}/conf/app.ini";
-          secretKey = "${cfg.customDir}/conf/secret_key";
-          oauth2JwtSecret = "${cfg.customDir}/conf/oauth2_jwt_secret";
-          oldLfsJwtSecret = "${cfg.customDir}/conf/jwt_secret"; # old file for LFS_JWT_SECRET
-          lfsJwtSecret = "${cfg.customDir}/conf/lfs_jwt_secret"; # new file for LFS_JWT_SECRET
-          internalToken = "${cfg.customDir}/conf/internal_token";
-          replaceSecretBin = "${pkgs.replace-secret}/bin/replace-secret";
-        in
         ''
-          # copy custom configuration and generate random secrets if needed
-          ${lib.optionalString (!cfg.useWizard) ''
+          ${optionalString (!cfg.useWizard) ''
             function forgejo_setup {
-              cp -f '${format.generate "app.ini" cfg.settings}' '${runConfig}'
-
-              if [ ! -s '${secretKey}' ]; then
-                  ${exe} generate secret SECRET_KEY > '${secretKey}'
-              fi
-
-              # Migrate LFS_JWT_SECRET filename
-              if [[ -s '${oldLfsJwtSecret}' && ! -s '${lfsJwtSecret}' ]]; then
-                  mv '${oldLfsJwtSecret}' '${lfsJwtSecret}'
-              fi
-
-              if [ ! -s '${oauth2JwtSecret}' ]; then
-                  ${exe} generate secret JWT_SECRET > '${oauth2JwtSecret}'
-              fi
-
-              ${optionalString cfg.lfs.enable ''
-              if [ ! -s '${lfsJwtSecret}' ]; then
-                  ${exe} generate secret LFS_JWT_SECRET > '${lfsJwtSecret}'
-              fi
-              ''}
-
-              if [ ! -s '${internalToken}' ]; then
-                  ${exe} generate secret INTERNAL_TOKEN > '${internalToken}'
-              fi
-
-              chmod u+w '${runConfig}'
-              ${replaceSecretBin} '#secretkey#' '${secretKey}' '${runConfig}'
-              ${replaceSecretBin} '#oauth2jwtsecret#' '${oauth2JwtSecret}' '${runConfig}'
-              ${replaceSecretBin} '#internaltoken#' '${internalToken}' '${runConfig}'
-
-              ${optionalString cfg.lfs.enable ''
-                ${replaceSecretBin} '#lfsjwtsecret#' '${lfsJwtSecret}' '${runConfig}'
-              ''}
-
-              ${optionalString (cfg.database.passwordFile != null) ''
-                ${replaceSecretBin} '#dbpass#' '${cfg.database.passwordFile}' '${runConfig}'
-              ''}
-
-              ${optionalString (cfg.mailerPasswordFile != null) ''
-                ${replaceSecretBin} '#mailerpass#' '${cfg.mailerPasswordFile}' '${runConfig}'
-              ''}
-              chmod u-w '${runConfig}'
+              config='${cfg.customDir}/conf/app.ini'
+              cp -f '${format.generate "app.ini" cfg.settings}' "$config"
+
+              chmod u+w "$config"
+              ${lib.getExe' cfg.package "environment-to-ini"} --config "$config"
+              chmod u-w "$config"
```

Comment:
this leads to the following error for me:

```
gitea[612]: 2024/06/02 18:04:21 ...es/setting/oauth2.go:151:loadOAuth2From() [F] save oauth2.JWT_SECRET failed: failed to save "/var/lib/forgejo/custom/conf/app.ini": open /var/lib/forgejo/custom/conf/app.ini: permission denied
```

i'm using this changed module with an existing forgejo installation data directory, _and_ i've changed the `oauth2.JWT_SECRET` from what was autogenerated.

Reply:
This looks to me like you messed up something in your config.

From what I can tell, Forgejo only does that when both `oauth2.JWT_SECRET` and `oauth2.JWT_SECRET_URI` are missing or are invalid (e.g. wrong format, wrong length).

Can you please share your NixOS config, more than just a single line of logs, and, if you can, either your secret or how you generated it?

Reply:
i can happily confirm my messing up :smile: 

the invalid token was generated using `openssl rand -hex 32` (command i found randomly online) and forgejo generated a new one as it wasn't suitable. i copied the new one into my secrets file now as a quick fix. eventually i'd like to look into how to pre-generate these secrets.

it didn't require a config change to fix my mistake. here it is anyway:

```nix    
    sops.secrets.FORGEJO_JWT_SECRET = {};
    sops.secrets.FORGEJO_INTERNAL_TOKEN = {};
    sops.secrets.FORGEJO_SECRET_KEY = {};

    services.forgejo = {
      enable = true;
      package = repoFlake.inputs.nixpkgs_forgejo.legacyPackages.${pkgs.system}.forgejo;
      settings = {
        service.DISABLE_REGISTRATION = true;
        server.HTTP_ADDR = "127.0.0.1";
        server.DISABLE_SSH = true;
        server.ROOT_URL = "https://forgejo.${domain}";
        server.HTTP_PORT = 3001;
      };
      secrets = {
        oauth2.JWT_SECRET = lib.mkForce config.sops.secrets.FORGEJO_JWT_SECRET.path;
        security.INTERNAL_TOKEN = lib.mkForce config.sops.secrets.FORGEJO_INTERNAL_TOKEN.path;
        security.SECRET_KEY = lib.mkForce config.sops.secrets.FORGEJO_SECRET_KEY.path;
      };
    };
```